pkg/reconcile/pipeline/context/impl.go

```diff
@@ -118,15 +121,44 @@ func (i *impl) Services() ([]pipeline.Service, error) {
 func (i *impl) Applications() ([]pipeline.Application, error) {
 	if i.applications == nil {
 		if ref := i.serviceBinding.Spec.Application; ref != nil {
-			gvr, err := i.typeLookup.ResourceForReferable(ref)
-			if err != nil {
-				return nil, err
+			if i.serviceBinding.Spec.Application.Name != "" &&
+				i.serviceBinding.Spec.Application.LabelSelector != nil &&
+				i.serviceBinding.Spec.Application.LabelSelector.MatchLabels != nil {
+				return nil, errs.New("Both label selector and name for application should be used")
 			}
-			u, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).Get(context.Background(), ref.Name, metav1.GetOptions{})
-			if err != nil {
-				return nil, err
+			if i.serviceBinding.Spec.Application.Name != "" {
+				gvr, err := i.typeLookup.ResourceForReferable(ref)
+				if err != nil {
+					return nil, err
+				}
+				u, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).Get(context.Background(), ref.Name, metav1.GetOptions{})
+				if err != nil {
+					return nil, err
+				}
+				i.applications = append(i.applications, &application{gvr: gvr, persistedResource: u, bindingPath: i.serviceBinding.Spec.Application.BindingPath})
+
+			}
+			if i.serviceBinding.Spec.Application.LabelSelector != nil && i.serviceBinding.Spec.Application.LabelSelector.MatchLabels != nil {
+				gvr, err := i.typeLookup.ResourceForReferable(ref)
+				matchLabels := i.serviceBinding.Spec.Application.LabelSelector.MatchLabels
+				opts := metav1.ListOptions{
+					LabelSelector: labels.Set(matchLabels).String(),
+				}
+
+				objList, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).List(context.Background(), opts)
+				if err != nil {
+					return nil, err
+				}
+
+				if len(objList.Items) == 0 {
+					return nil, errs.New("application not found")
```

Comment:
the contract of this method is to return empty array, not array if not applications are not found, see the doc: https://github.com/redhat-developer/service-binding-operator/blob/d680987ea0f2869557ada3f3dcc621981cb1d427/pkg/reconcile/pipeline/api.go#L102

Reply:
Returning `nil` returns an empty slice of `Application` objects.
See this example: https://play.golang.org/p/QvQLT9CH3Fg

Reply:
Thanks, nice. However, according to the contract, we should not return  error in this place, just empty slice.

Reply:
Fixed and added a test case for the same: 588535ccd6305f26c5dfe8e2ac33909a1d313ad9